### ISSUE TITLE
Example programs hang looking for Tau Camera hardware when here is an additional, non-responsive serial port on the computer

### ISSUE THREAD

**Issue by greenbreakfast** [OnionIoT/tau-lidar-camera#17]
### Python Library versions
> Run `pip list | grep TauLidar` in the command-line and paste the output. Include some details on your system as well

TauLidarCamera 0.0.4
TauLidarCommon 0.0.2

Running on a Mac

### Camera Info
> If possible, paste the output of `camera.info()` from Python (after camera has been successfully opened)

```
    model:      4.0
    firmware:   3.3
    uid:        450225
    resolution: 160x60
    port:       /dev/tty.usbmodem00000000001A1
```

### Describe the bug

Example programs hang on `Looking for connected Tau LiDAR Camera hardware ...` when there is an additional, non-responsive serial port on the computer

### Steps to Reproduce

* There are several serial ports on my Mac:
  ![image](https://user-images.githubusercontent.com/6757320/126837126-15523251-cfc1-4dde-b42b-78981f126877.png)
  * The Tau Camera is on `/dev/tty.usbmodem00000000001A1`
  * Not really sure where `/dev/tty.usbmodem104NTZNH11332` comes from or what it is
* Run `distance.py` example program
* Hangs on `Looking for connected Tau LiDAR Camera hardware ...` message indefinitely
  ![image](https://user-images.githubusercontent.com/6757320/126837263-8dc0a3c9-ce41-4e57-932e-dbe714ce68f6.png)

### What is expected?

Expecting the example program to:
1. scan the serial ports
2. establish connection to the connected Tau camera
3. initialize the camera
4. start sending depth frames to the computer

### What is actually happening?

The program hangs on the step where it's `Looking for connected Tau LiDAR Camera hardware`
Cannot be killed with `ctrl+c`, can only be stopped by killing the process

### Any additional comments?

**The program seems to get stuck in the `ports = Camera.scan()` call**

I've updated the example programs to have a command-line option to specify the serial port (see commit a86c52f4a193aa918f50561c7080237fa2079047). 
The example programs work fine when the serial port is specified and the `scan` function is avoided, so the Camera itself is not the issue:

![image](https://user-images.githubusercontent.com/6757320/126838020-46fd7fff-7525-44ab-b964-9b866c697df0.png)

**Comment by greenbreakfast** [comment]
@vsemi could you please look into this? 
Perhaps we can add a timeout when attempting to connect to each port in the `scan` function?

**Comment by vsemi** [comment]
I'm not able reproduce this issue, but by reading the code, that could be problem of waiting for (problematic) serial port response, as normally there is just very limited number of available serial ports to be looped, the hanging only possible happens in a specific loop while trying communicating to a (problematic) serial port which somehow does not respond and be more weird that it hanging in pyserial library, and we do have one mini-second timeout setting for pyserial, not sure what happens it failed to give up in such case after one mini-second. I do not have a Mac machine, not sure if it happens on Mac only, but I think it might happens only when there is a fake serial port or problematic serial port. Suggested user to use device manager of the computer to investigate that problem serial port, to fix or disable it.

**Comment by greenbreakfast** [comment]
@vsemi It's not a feasible solution to have users disable serial ports on their machines. Most serial ports have some other purpose on the machine. 
For example `/dev/cu.Bluetooth-Incoming-Port` on Macs has something to do with Bluetooth. Disabling it is not an option since that would likely cause issues with how bluetooth work.

On top of that, our goal is to make the Tau Camera easy to use out of the box and this issue is definitely fixable in our code.

---

I took a deeper dive and found where the issue lies. Here is a trace of what happens with the `/dev/cu.Bluetooth-Incoming-Port` serial port on my machine:

1. In the `scan` function we loop through all available serial ports and call `getIdentification` on each of them
1. In `getIdentification`, we `_sendCommandWithoutData` to the serial port, which then uses `_sendCommand`
1. In `_sendCommand`, the `_write` function works properly, but it's the `_read` function where the problem happens
1. In `_read`, the issue happens in the `while` loop:
    1. `_ser.read` works properly - hence no timeout - but what's returned is a `b''`
    2. This turns into a 0-length byte-array and gets added to `array`, which **doesn't change the length of `array`**
    3. Then, when the while loop re-evaluates the loop condition, `len(array)` is always `0` resulting in an **infinite loop**

Same thing happens with the `/dev/tty.usbmodem104NTZNH11332` serial port on my machine.

**Can you please look into updating the `Communication._read` function to have some sort of infinite loop protection?**

**Comment by vsemi** [comment]
@greenbreakfast line 156 in Communication._read, limited the maximum attempts to be 100, is not an infinite loop:

if count > 100 : break

Could you help to debug how it does not beak the loop over there?

**Comment by greenbreakfast** [comment]
@vsemi looks like the max attempt limit was taken out in your fix for #14 (commit 3e0036d31f49cde28154f0d2ff96468c27307bbc) 

There's a `count` variable but nothing to break the loop: https://github.com/OnionIoT/tau-lidar-camera/blob/master/TauLidarCamera/communication.py#L136

**Comment by vsemi** [comment]
@greenbreakfast my bad, how comes I made such a mistake, in my vsemi repo where I investigate the issue I still have that 2 lines, must be a mistake when I copy and paste that block over. Added back in branch issue_17. Let's see if it can break the loop on your Mac.

**Comment by greenbreakfast** [comment]
@vsemi thanks, the update in `issue_17` fixes the bug I was seeing. I'll merge into master and make a new release.

How was the 100 loop limit chosen? If it's an arbitrary number, how can we be sure that's enough loop attempts?

**Comment by vsemi** [comment]
@greenbreakfast  1) read timeout setting is one mini-second, 100 attempts means 100 mini-seconds, the slowest frame-rate (10 fps) expected in very low edge devices. 1) tests on different platforms, includes edge device (Omega), observed normally it can be done in attempts.

**Comment by greenbreakfast** [comment]
@vsemi ok, sounds reasonable. Thanks again!

Fix released in v0.0.5
